### ISSUE TITLE
[Feature] MLX Graph Runner for Apple Silicon Backend

### ISSUE THREAD

**Issue by shvmjndl** [sgl-project/sglang#19145]
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

This issue tracks the design and implementation of `MLXGraphRunner` — the graph runner component of the MLX backend for Apple Silicon support (part of the roadmap in #17846).

#### Why a Graph Runner is Needed

In SGLang, every device backend requires a **graph runner** — the component responsible for executing the model's forward pass during inference. The `ModelRunner` dispatches to the appropriate graph runner based on the device type (`model_runner.py:2149-2156`):

```python
graph_runners = defaultdict(
    lambda: CudaGraphRunner,
    {
        "cpu": CPUGraphRunner,
        "npu": NPUGraphRunner,
    },
)
self.graph_runner = graph_runners[self.device](self)
```

Without an MLX graph runner, the MLX backend cannot execute inference. This is a **required MVP component**.

#### Why MLX Needs Its Own Runner (Not Just CudaGraphRunner)

The existing `CudaGraphRunner` relies on CUDA-specific APIs (`torch.cuda.CUDAGraph`, `graph.capture()`, `graph.replay()`, CUDA streams, GPU memory pools) that do not exist on Apple Silicon. The `CPUGraphRunner` uses `torch.compile` with CPU-specific fake ops and is also not directly reusable.

MLX's architecture is fundamentally different:
- **Unified memory**: No CPU↔GPU transfer overhead, eliminating the need for CUDA-style graph capture/replay
- **Lazy evaluation**: Computation is deferred until `mx.eval()` is called
- **JIT compilation**: `mx.compile()` provides kernel fusion and optimization, replacing CUDA graph's role of batching kernel launches

This means the MLX graph runner can be significantly simpler — a thin wrapper that leverages `mx.compile()` rather than the complex capture/replay machinery of CUDA graphs.

---

#### File Location

```
python/sglang/srt/hardware_backend/mlx/mlx_graph_runner.py
```

#### Required Interface

The graph runner must implement the contract expected by `ModelRunner`:

| Method | Purpose |
|--------|---------|
| `__init__(self, model_runner)` | Initialize runner, set up compiled forward functions |
| `can_run(self, forward_batch) -> bool` | Check if this runner can handle the given batch |
| `replay(self, forward_batch, skip_attn_backend_init, pp_proxy_tensors) -> LogitsProcessorOutput \| PPProxyTensors` | Execute the model forward pass |

#### Architecture

```
MLXGraphRunner (standalone class, does NOT inherit CudaGraphRunner)
│
├── __init__(model_runner)
│   ├── Store model_runner reference and config
│   ├── Set up mx.compile() wrapped forward function
│   └── Assert unsupported features (LoRA, speculative, PP, encoder-decoder, etc.)
│
├── can_run(forward_batch) -> bool
│   ├── Check batch size within limits
│   └── Check capture_hidden_mode compatibility
│
└── replay(forward_batch, ...) -> LogitsProcessorOutput | PPProxyTensors
    ├── Initialize attention backend metadata
    ├── Execute compiled forward pass
    ├── mx.eval() to force lazy computation
    └── Return LogitsProcessorOutput
```

#### Key Design Decisions

**1. Standalone class (like CPUGraphRunner), not subclassing CudaGraphRunner**

The NPU runner subclasses `CudaGraphRunner` because NPU graphs work similarly to CUDA graphs. MLX has no graph capture/replay concept, so inheriting would bring unnecessary complexity and unused CUDA-specific code paths.

**2. `mx.compile()` for optimization**

MLX's JIT compiler (`mx.compile()`) fuses operations and optimizes Metal shader dispatch. This serves the same role as CUDA graphs (reducing kernel launch overhead) but through compilation rather than capture/replay:

```python
def __init__(self, model_runner):
    self.compiled_forward = mx.compile(self._forward_impl)
```

**3. No batch-size-specific graph capture**

CUDA graphs require capturing a separate graph per batch size. `mx.compile()` handles dynamic shapes natively (or can be called with specific input shapes for optimization), so we don't need the `capture_bs` / `graphs[bs]` pattern.

**4. Tensor bridge between PyTorch and MLX**

Since SGLang's core uses PyTorch tensors but MLX uses its own array type, the runner needs conversion at the boundary:

```python
# PyTorch → MLX (input)
input_ids_mlx = mx.array(input_ids.cpu().numpy())

# MLX → PyTorch (output)
mx.eval(output)
logits_torch = torch.from_numpy(np.array(output))
```

This is an MVP approach. Future optimization can minimize or eliminate these conversions.

**5. MVP scope limitations**

The initial implementation will assert that these features are not enabled:
- LoRA
- Speculative decoding (EAGLE, ngram)
- Pipeline parallelism (PP > 1)
- Data parallelism (DP > 1)
- Tensor parallelism (TP > 1)
- Two-batch overlap
- Encoder-decoder models

These mirror the same limitations as `CPUGraphRunner` and can be relaxed in later phases.

#### Implementation Sketch

```python
class MLXGraphRunner:
    """Graph runner for Apple Silicon using MLX."""

    def __init__(self, model_runner: "ModelRunner"):
        self.model_runner = model_runner
        self.device = model_runner.device
        
        # Validate unsupported features
        assert not model_runner.server_args.enable_lora
        assert model_runner.server_args.dp_size == 1
        assert model_runner.server_args.pp_size == 1
        assert model_runner.spec_algorithm == SpeculativeAlgorithm.NONE
        
        # Set up capture modes
        self.capture_hidden_mode = CaptureHiddenMode.NULL
        if model_runner.server_args.enable_return_hidden_states:
            self.capture_hidden_mode = CaptureHiddenMode.FULL
        
        # Compile model forward with MLX JIT
        self.compiled_forward = mx.compile(self._forward_wrapper)
    
    def can_run(self, forward_batch: ForwardBatch) -> bool:
        requested_mode = max(
            forward_batch.capture_hidden_mode,
            getattr(forward_batch.spec_info, "capture_hidden_mode", CaptureHiddenMode.NULL),
        )
        return (requested_mode == CaptureHiddenMode.NULL 
                or requested_mode == self.capture_hidden_mode)
    
    def replay(self, forward_batch, skip_attn_backend_init=False, pp_proxy_tensors=None):
        if not skip_attn_backend_init:
            self.model_runner.attn_backend.init_forward_metadata(forward_batch)
        
        output = self.compiled_forward(
            forward_batch.input_ids,
            forward_batch.positions,
            forward_batch,
        )
        return output
```

#### Registration Change

In `model_runner.py`:

```python
from sglang.srt.hardware_backend.mlx.mlx_graph_runner import MLXGraphRunner

graph_runners = defaultdict(
    lambda: CudaGraphRunner,
    {
        "cpu": CPUGraphRunner,
        "npu": NPUGraphRunner,
        "mlx": MLXGraphRunner,  # ← new
    },
)
```

---

### Dependencies

- MLX attention backend (`mlx_attention.py`) — the graph runner calls `attn_backend.init_forward_metadata()`
- MLX model loader (`mlx_model_loader.py`) — the model must be loaded before the graph runner initializes
- Device config update — `"mlx"` must be added to the valid device list in `device_config.py`

### Testing Plan

1. Unit test: Instantiate `MLXGraphRunner` with a mock `ModelRunner` and verify `can_run()` / `replay()` interface
2. Integration test: End-to-end inference with a small model (e.g., `mlx-community/TinyLlama-1.1B-Chat-v1.0-4bit`)
3. Correctness: Compare output logits against reference MLX-LM implementation

### Related

- Parent issue: #17846 (MLX Backend Implementation for SGLang)
- Reference implementations: `cpu_graph_runner.py`, `npu_graph_runner.py`
- MLX compile docs: https://ml-explore.github.io/mlx/build/html/usage/compile.html

**Comment by shvmjndl** [comment]
Closing — will re-open with a revised design plan.
